pyro/distributions/diag_normal_mixture_shared_cov.py

```diff
@@ -0,0 +1,139 @@
+
+    [1] Pathwise Derivatives for Multivariate Distributions, Martin Jankowiak & Theofanis Karaletsos.
+        arXiv:1806.01856
+
+    :param torch.Tensor locs: K x D mean matrix
+    :param torch.Tensor scale: shared D-dimensional scale vector
+    :param torch.Tensor logits: K-dimensional vector of softmax logits
+    """
+    has_rsample = True
+    arg_constraints = {"locs": constraints.real, "scale": constraints.positive,
+                       "logits": constraints.real}
+
+    def __init__(self, locs, scale, logits):
+        self.batch_mode = (locs.dim() > 2)
+        assert(self.batch_mode or locs.dim() == 2), \
+            "The locs parameter in MixtureOfDiagNormals should be K x D dimensional (or ... x B x K x D in batch mode)"
+        if not self.batch_mode:
+            assert(scale.dim() == 1), "The scale parameter in MixtureOfDiagNormals should be D dimensional"
+            assert(logits.dim() == 1), "The logits parameter in MixtureOfDiagNormals should be K dimensional"
+            assert(logits.size(0) == locs.size(0))
+            batch_shape = ()
+        else:
+            assert(scale.dim() > 1), "The scale parameter in MixtureOfDiagNormals should be ... x B x D dimensional"
+            assert(logits.dim() > 1), "The logits parameter in MixtureOfDiagNormals should be ... x B x K dimensional"
+            assert(logits.size(-1) == locs.size(-2))
+            batch_shape = tuple(locs.shape[:-2])
+        self.locs = locs
+        self.scale = scale
+        self.logits = logits
+        self.dim = locs.size(-1)
+        self.categorical = Categorical(logits=logits)
+        self.probs = self.categorical.probs
+        super(MixtureOfDiagNormalsSharedCovariance, self).__init__(batch_shape=batch_shape, event_shape=(self.dim,))
+
+    def log_prob(self, value):
+        scale = self.scale.unsqueeze(-2) if self.batch_mode else self.scale
+        epsilon = (value.unsqueeze(-2) - self.locs) / scale  # L B K D
+        eps_sqr = 0.5 * torch.pow(epsilon, 2.0).sum(-1)  # L B K
+        eps_sqr_min = torch.min(eps_sqr, -1)[0]  # L B K
+        result = self.probs * torch.exp(-eps_sqr + eps_sqr_min.unsqueeze(-1))  # L B K
+        result = torch.log(result.sum(-1))  # L B
+        result -= 0.5 * math.log(2.0 * math.pi) * float(self.dim)
+        result -= torch.log(self.scale).sum(-1)
+        result -= eps_sqr_min
+        return result
+
+    def rsample(self, sample_shape=torch.Size()):
+        which = self.categorical.sample(sample_shape)
+        return _MixDiagNormalSharedCovarianceSample.apply(self.locs, self.scale, self.logits, self.probs,
+                                                          which, sample_shape + self.scale.shape)
+
+
+class _MixDiagNormalSharedCovarianceSample(Function):
+    @staticmethod
+    def forward(ctx, locs, scale, logits, pis, which, noise_shape):
+        dim = scale.size(-1)
+        white = locs.new(noise_shape).normal_()
+        n_unsqueezes = locs.dim() - which.dim()
```

Comment:
Can you explain `n_unsqueezes` with a comment? I'm having trouble seeing why it should ever be anything but 1. In particular, if `len(sample_shape) > 0`, I still think you should `.unsqueeze(-1)` exactly once, and additionally `locs.expand(which.shape + (dim,))`.

Reply:
i can't recall exactly but i needed this to get gather to work properly and produce correct samples. 

if you run the test you will find that `n_squeezes` takes values 2/1 for `batch_mode` = True/False